### PR TITLE
Refactor Config Package

### PR DESCRIPTION
**Pull request by eliasjf**
This PR refactors the config package to remove the `idp.LdapIDP` field. Part of those changes include reworking how a `tls.Config` is generated. This refactor is intended to setup the upcoming CONNECTOR work.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -96,28 +103,85 @@ func (cm *ConfigManager) ReadConfig(filename string) (*Config, error) {
...
+		}
+	}
+
+	cm.Logger.Log("level", "info", "msg", "loaded "+filename, "version", c.Version)
+	return &c, nil
+}
+
+// processOutboundTLS builds a *tls.Config for outbound connections to remote services. This *tls.Config can be used by
+// connectors to authenticate against a target IDP.
+func processOutboundTLS(logger log.Logger, fs afero.Fs, conf TLS) (*tls.Config, error) {
```
Comment (eliasjf) [inline pkg/config/config.go]: This method was ported over from `ldap_config.go`. I changed the name and slightly changed the signature.

File: pkg/config/config.go [hunk 2]
```diff
@@ -27,14 +25,23 @@ var loadX509KeyPair = func(certFile, keyFile string) (tls.Certificate, error) {
...
+	URL           string
+	TLS           TLS `yaml:"tls"`
 	TLSConfig     *tls.Config
 }
 
 // TLS reprents TLS-related configuration options for the Maverics service.
 type TLS struct {
+	// InboundTLS is a flag to specify whether the TLS config is intended for inbound connections. Inbound TLS is used
+	// for client authentication while outbound TLS is used to authenticate against a remote service.
+	InboundTLS bool `yaml:"inboundTLS"`
```
Comment (eliasjf) [inline pkg/config/config.go]: I expect we can ultimately remove this flag. I didn't want to get bogged down with TLS specific work and saw this as the cleanest way to minimize changes.

File: pkg/config/config.go [hunk 3]
```diff
@@ -27,14 +25,23 @@ var loadX509KeyPair = func(certFile, keyFile string) (tls.Certificate, error) {
 // Config represents configuration options for the Maverics service.
 type Config struct {
 	Version       string
-	ListenAddress string      `yaml:"listenAddress"`
-	LDAP          idp.LdapIDP `yaml:"ldap"`
-	TLS           TLS         `yaml:"tls"`
+	ListenAddress string `yaml:"listenAddress"`
+	Name          string
```
Comment (topherMarie) [inline pkg/config/config.go]: What is the Name for in here?

File: pkg/config/config.go (hunk 2 above)
Comment (rmulley) [reply pkg/config/config.go]: Agree, I wonder if the structure even needs this field and can just populate what is provided. The configuration could dictate whether it's inbound or outbound.

File: pkg/config/config.go [hunk 4]
```diff
@@ -96,28 +103,85 @@ func (cm *ConfigManager) ReadConfig(filename string) (*Config, error) {
...
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to load X509 key pair")
+		}
+
+		certs = append(certs, cert)
+	}
+
+	// Only attempt to load root CAs if configuration data was provided.
+	var rootCAs *x509.CertPool
+	if conf.CAFile != "" {
```
Comment (rmulley) [inline pkg/config/config.go]: I wonder if this block could be combined with `buildClientCACertPool`, provided the method was made slightly more generic.

Comment (rmulley) [reply pkg/config/config.go]: We don't have to do this now if we don't want to.

**Comment by eliasjf** [comment]
@topherMarie, @rmulley, I've updated this PR to:
1. Remove the changes I had previously made to the LDAP connector and corresponding example. This will help avoid a nasty merge conflict since Robert is working against the LDAP connector.
2. Reworked the config package to add the concept of `Connectors` and added a "dummy" connector and corresponding example to show this in action.

You'll notice that the tests are currently failing. This is because the config changes break both `examples/ldap` and `cmd/maverics`. I of course plan to fix these programs but I think it will be useful to get some interim feedback and to possibly make the fixes in another PR.

**Pull request closed without merging**